# 00 · Start Here

Companion notebook for the [lesson](https://github.com/younespuri/visual-machine-learning-for-beginners/blob/main/lessons/00-start-here/README.md). Run the cells from top to bottom with **Shift + Enter**.

### Step 1 · NumPy arrays: lists that do math

In [1]:
import numpy as np

a = np.array([50, 60, 70, 85, 100])
print(a * 2)        # math on every element at once
print(a.mean())
print(a.shape)      # 5 elements in one dimension
# [100 120 140 170 200]
# 73.0
# (5,)

table = a.reshape(-1, 1)   # the same numbers as one column: 5 rows, 1 column
print(table.shape)
# (5, 1)

[100 120 140 170 200]
73.0
(5,)
(5, 1)


- A **NumPy array** is like a Python list that can do math on all its numbers at once, fast.
- `shape` tells you the size of each dimension. `(5,)` is a flat list of 5; `(5, 1)` is a table with 5 rows and 1 column.
- `reshape(-1, 1)` turns a flat list into a one-column table. You'll use it in [lesson 02](../02-linear-regression/README.md), because models want their inputs as tables.

### Step 2 · pandas DataFrames: tables with column names

In [2]:
import pandas as pd

df = pd.DataFrame({
    "area":  [50, 60, 70, 85, 100, 120, 140, 65, 95, 110],
    "rooms": [1, 2, 2, 2, 3, 3, 4, 2, 3, 3],
    "age":   [30, 25, 15, 12, 8, 20, 5, 18, 10, 7],
    "price": [540, 560, 780, 790, 1040, 1080, 1340, 700, 960, 1150],
})
print(df.head(3))
print(df.shape)
#    area  rooms  age  price
# 0    50      1   30    540
# 1    60      2   25    560
# 2    70      2   15    780
# (10, 4)

   area  rooms  age  price
0    50      1   30    540
1    60      2   25    560
2    70      2   15    780
(10, 4)


- A **DataFrame** is a table with named columns, like a spreadsheet in Python.
- `df.head(3)` shows the first 3 rows; `df.shape` says 10 rows and 4 columns.

### Step 3 · Choose X and y

In [3]:
X = df[["area", "rooms", "age"]]   # the features: a table
y = df["price"]                    # the label: one column
print(X.shape, y.shape)
# (10, 3) (10,)

(10, 3) (10,)


- Double brackets `[[...]]` pick several columns and give back a table; single brackets pick one column.
- X has 10 rows and 3 features. y has 10 answers, one per row.

### Step 4 · Split into training and test sets

In [4]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
print("train:", X_train.shape, " test:", X_test.shape)
print(X_test)
# train: (8, 3)  test: (2, 3)
#    area  rooms  age
# 8    95      3   10
# 1    60      2   25

train: (8, 3)  test: (2, 3)
   area  rooms  age
8    95      3   10
1    60      2   25


- `test_size=0.2` keeps 20% of the rows for testing: 2 of our 10 houses.
- `train_test_split` shuffles before splitting, which is why the test rows are 8 and 1, not the last two.
- `random_state=42` makes the shuffle repeatable, so you get the same split as this lesson. Any number works; 42 is just a popular choice.

### Step 5 · Your first model, start to finish

In [5]:
from sklearn.datasets import load_iris
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score

# 1-2. collect the data, then choose X and y
iris = load_iris(as_frame=True)
X = iris.data
y = iris.target
print(X.shape)
print(iris.target_names)
# (150, 4)
# ['setosa' 'versicolor' 'virginica']

# 3. split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42, stratify=y)

# 4. fit
model = DecisionTreeClassifier(random_state=42)
model.fit(X_train, y_train)

# 5. predict
y_pred = model.predict(X_test)
print("guesses:", iris.target_names[y_pred[:4]])
print("answers:", iris.target_names[y_test.values[:4]])
# guesses: ['setosa' 'virginica' 'versicolor' 'versicolor']
# answers: ['setosa' 'virginica' 'versicolor' 'versicolor']

# 6. evaluate
print("accuracy:", round(accuracy_score(y_test, y_pred), 3))
# accuracy: 0.933

(150, 4)
['setosa' 'versicolor' 'virginica']
guesses: ['setosa' 'virginica' 'versicolor' 'versicolor']
answers: ['setosa' 'virginica' 'versicolor' 'versicolor']
accuracy: 0.933


- The model got **93%** of the 30 test flowers right: flowers it had never seen during training.
- `stratify=y` keeps the same mix of the three species in both sets, so the test isn't accidentally lopsided.
- Notice the pattern: create a model, `fit`, `predict`, evaluate. It stays the same for every model in this course. Only the one line that creates the model changes.